core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java

```diff
@@ -250,10 +250,11 @@ public boolean equals(Object o) {
 
         ByteSizeValue sizeValue = (ByteSizeValue) o;
 
-        if (size != sizeValue.size) return false;
-        if (sizeUnit != sizeValue.sizeUnit) return false;
-
-        return true;
+        if (sizeUnit == sizeValue.sizeUnit) {
+            return size == sizeValue.size;
+        } else {
+            return bytes() == sizeValue.bytes();
+        }
```

Comment:
Conversion should be very cheap, so I would prefer to always convert to avoid a branch.